### PR TITLE
Add Ethereum secp keys to validators

### PR DESCRIPTION
**Pull request by sug0**
Finalizes the implementation of the 8th item of issue no. https://github.com/anoma/namada/issues/242

This PR adds Ethereum secp keys to genesis validators, and also finalizes the validation code for valset upd vote extensions.

We integrate changes from https://github.com/anoma/namada/pull/371

---

**Missing in this PR:** Proof-of-stake VP changes, centered around verifying Ethereum secp keys in storage. This can be addressed in a future PR.

### CODE REVIEW

**Comment by sug0** [comment]
still missing some VP stuff, will make this a draft PR again

**Comment by sug0** [comment]
@james-chf feel free to leave a review, if you have time. what's blocking this PR right now are some changes to VPs, which will unify the native and wasm interfaces. after that is more or less done, in `main`, me/tomas/brent can implement the remaining stuff for the ledger secp keys pr https://github.com/anoma/namada/pull/371, which is being tracked here as well. the remaining changes will be around the PoS VP, to verify eth secp keys.

File: apps/src/lib/config/genesis.rs
Comment (tzemanovic) [reply apps/src/lib/config/genesis.rs:325]: ah, bad copy paste, nice catch!

File: shared/src/types/key/secp256k1.rs
Comment (tzemanovic) [reply shared/src/types/key/secp256k1.rs:424]: unfortunately, "secp256k1-sign-verify" feature bloats wasm build by huge amount even if it's not used, so we need to switch it off for it

Comment (batconjurer) [reply shared/src/types/key/secp256k1.rs:424]: :cry:  I see

Comment (sug0) [reply shared/src/types/key/secp256k1.rs:424]: tomas told me we feature gate some secp key things so that we do not bloat wasm binaries too much

**Comment by james-chf** [comment]
This PR has code from another PR + some new Ethereum-bridge specific logic, this would be easier to review in separate stacked PRs (though not worth splitting out now)

I can't test this locally yet as I need to update my way of generating a devchain to work with the new Tendermint v0.38.x, but I'll do that soon. Then I can test this altogether (maybe after it has merged to `eth-bridge-integration`), still have a couple more things to look at on this PR

File: shared/src/ledger/pos/vp.rs
Comment (sug0) [reply shared/src/ledger/pos/vp.rs:435]: it can happen if we are a fullnode, for instance

File: shared/src/ledger/pos/storage.rs
Comment (sug0) [reply shared/src/ledger/pos/storage.rs:614]: yeah this is more of a general issue outside of this PR, idk if @tzemanovic wants to pitch in on this

File: shared/src/ledger/pos/mod.rs
Comment (sug0) [reply shared/src/ledger/pos/mod.rs:112]: pretty sure this is the same reason why we have a `RefTo` trait lying around

File: apps/src/lib/wallet/store.rs
Comment (sug0) [reply apps/src/lib/wallet/store.rs:292]: yeah agreed. it forces you to return `Option` or `Return` types, which makes APIs more convoluted to use in the end.

File: apps/src/lib/node/ledger/shell/queries.rs
Comment (sug0) [reply apps/src/lib/node/ledger/shell/queries.rs:630]: same issue as above, it can return `None` if we are a fullnode.

Comment (sug0) [reply apps/src/lib/node/ledger/shell/queries.rs:630]: like we discussed in our call, the `Shell<...>` should take an extra parameter`NodeTag`  to distinguish between different node kinds, and only implement methods like `prepare_proposal` e.g. if we are a `NodeTag = Validator`. this would also obviate the need to error out from methods like `get_ethbridge_from_namada_addr`, unless we actually got some storage reading error.

File: shared/src/ledger/pos/mod.rs
Comment (james-chf) [reply shared/src/ledger/pos/mod.rs:112]: Ah interesting, it'd be nice if we could use `TryInto` but I haven't see what the code looks like with using that, if it means lifetimes everywhere

File: apps/src/lib/node/ledger/shell/queries.rs
Comment (james-chf) [reply apps/src/lib/node/ledger/shell/queries.rs:630]: Yeah that sounds like a good approach, have opened https://github.com/anoma/namada/issues/521

File: shared/src/ledger/pos/vp.rs
Comment (james-chf) [reply shared/src/ledger/pos/vp.rs:435]: Even for a full node, we are reading the validator's key from blockchain state (`self.ctx.read_pre`) when executing this VP, so these two funcs should work the same for full nodes and validators? I guess returning an `Option` here follows the pattern of other methods here for the time being

**Review by james-chf — APPROVED** [review]
lgtm to merge!

File: shared/src/ledger/pos/storage.rs
Comment (tzemanovic) [reply shared/src/ledger/pos/storage.rs:614]: In general, yes agree that it'd be better. We didn't worry about this too much for now in here as this is inside `PosBase` and these methods are only used by the protocol to init chain. If we return an error from, it would still crash further up the stack - the API of PoS crate could do some bigger refactors in general, but for now it's not a high priority

File: shared/src/ledger/pos/vp.rs
Comment (tzemanovic) [reply shared/src/ledger/pos/vp.rs:435]: at type level `key: &Self::Address` can be any address, so this would be `None` if the given address is not a validator address. If we know we're calling it with validator addresses only, we can expect that those are always `Some`

File: shared/src/ledger/pos/mod.rs
Comment (tzemanovic) [reply shared/src/ledger/pos/mod.rs:112]: yeah, we had to specify pretty funky bounds on the PoS traits that use an associated types for the public key to get it to work, so this was actually lot cleaner

Comment (tzemanovic) [reply shared/src/ledger/pos/mod.rs:112]: still, same as the other thing, the way PoS traits are implemented could use a bigger refactor that would alleviate this

**Review by batconjurer — APPROVED** [review]

**Comment by sug0** [comment]
pls update wasm

**Pull request merged**
